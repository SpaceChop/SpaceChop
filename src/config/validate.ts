import console from '../lib/console';
import validate from '../types/validate';

let valid = true;

export default (config, throwError = false) => {
  try {
    validate(config);
    if (!valid) {
      valid = true;
      console.info('Config is valid again');
    }
    return true;
  } catch (err) {
    valid = false;
    console.error('\nThere is an error in /config.yml:');
    console.error('--------------------------------');
    console.error(err.message);
    console.error('--------------------------------');
    if (throwError) {
      throw new Error(err);
    }
  }
  return false;
};
