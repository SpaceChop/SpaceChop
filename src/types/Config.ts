import * as t from 'runtypes';
import validateParams from '../lib/validateParams';
import Paths from './Paths';
import Presets from './Presets';
import Sources from './Sources';
import Storage from './Storage';

const Config = t.Record({
  paths: Paths,
  sources: Sources,
  presets: Presets,
}).And(t.Partial({
  storage: Storage,
  disableChunkedEncoding: t.Boolean,
})).withConstraint(validateParams);

export type Config = t.Static<typeof Config>;
export default Config;
