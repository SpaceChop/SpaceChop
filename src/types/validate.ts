import Config from './Config';

export default (config) => {
  try {
    Config.check(config);
  } catch (err) {
    throw new Error(`${err.key ? `${err.key}: ` : ''}${err.message}`);
  }
};
