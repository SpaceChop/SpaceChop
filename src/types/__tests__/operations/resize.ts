import { defaultConfig } from '.';
import validate from '../..';
import { Config } from '../../Config';

describe('should validate resize config', () => {
  describe('should throw on invalid config', () => {
    it('require width or height', () => {
      const config = {
        ...defaultConfig,
        presets: {
          t_200: {
            steps: [{
              $resize: {
              },
            }],
          },
        },
      };
      expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    });

    it('require valid width and height', () => {
      let config = {
        ...defaultConfig,
        presets: {
          t_200: {
            steps: [{
              $resize: {
                width: -100,
                height: 100,
              },
            }],
          },
        },
      };
      expect(() => validate(config)).toThrowErrorMatchingSnapshot();

      config = {
        ...defaultConfig,
        presets: {
          t_200: {
            steps: [{
              $resize: {
                width: 100,
                height: -100,
              },
            }],
          },
        },
      };
      expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    });
  });

  describe('should now throw on valid config', () => {
    it('width & height', () => {
      const config: Config = {
        ...defaultConfig,
        presets: {
          t_200: {
            steps: [{
              $resize: {
                width: 100,
                height: 100,
              },
            }],
          },
        },
      };
      expect(() => validate(config)).not.toThrow();
    });
  });
});
