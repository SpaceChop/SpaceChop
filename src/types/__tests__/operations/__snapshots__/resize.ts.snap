// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should validate operations no double operations in step 1`] = `"presets.t_200.steps.[0]: please only one operation per step"`;

exports[`should validate resize config should throw on invalid config require valid width and height 1`] = `"presets.t_200.steps.[0].$resize.width: Expected number | { from_path: string; }, but was number"`;

exports[`should validate resize config should throw on invalid config require valid width and height 2`] = `"presets.t_200.steps.[0].$resize.height: Expected number | { from_path: string; }, but was number"`;

exports[`should validate resize config should throw on invalid config require width or height 1`] = `"presets.t_200.steps.[0].$resize.width: Expected number | { from_path: string; }, but was undefined"`;
